dispatcher/main.go

```diff
@@ -0,0 +1,480 @@
+					}
+				}
+			}
+		}
+
+		if _, ok := statuses["initialized"]; !ok {
+			if _, ok := statuses["scheduled"]; !ok {
+				info(fmt.Sprintf("Run %d is finished", run.ID))
+				report(run)
+				os.Exit(0)
+			} else if statuses["scheduled"]-len(lostTask) == 0 {
+				info(fmt.Sprintf("Run %d is finished despite %d lost tasks.", run.ID, len(lostTask)))
+				report(run)
+				os.Exit(0)
+			}
+		}
+
+		time.Sleep(time.Second * 30)
+	}
+}
+
+func report(run *models.Run) {
+	info("Sending report...")
+	if email, ok := run.Settings[common.KeyUserEmail]; ok {
+		content := make(map[string]string)
+		content["run_id"] = strconv.Itoa(run.ID)
+		content["receivers"] = email.(string)
+
+		body, err := json.Marshal(content)
+		if err != nil {
+			info("Fail to marshel JSON during request sending email.")
+			return
+		}
+
+		info(string(body))
+		req, err := http.NewRequest("POST", "http://a01-svc-email-service/report", bytes.NewBuffer(body))
+		if err != nil {
+			info("Fail to create request to requesting email.")
+			return
+		}
+		req.Header.Set("Content-Type", "application/json")
+
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			info("Fail to send request to email service.")
+			return
+		}
+
+		respContent, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			info("Fail to read response body.")
+			return
+		}
+
+		info(fmt.Sprintf("Finish sending report: %s", string(respContent)))
+	} else {
+		info("Skip sending report")
+	}
+}
+
+func queryTests(run *models.Run) []models.TaskSetting {
+	info(fmt.Sprintf("Expecting script %s.", indexScript))
+	content, err := exec.Command(indexScript).Output()
+	if err != nil {
+		panic(err.Error())
+	}
+
+	var input []models.TaskSetting
+	err = json.Unmarshal(content, &input)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	if query, ok := run.Settings[common.KeyTestQuery]; ok {
+		info(fmt.Sprintf("Query string is '%s'", query))
+		result := make([]models.TaskSetting, 0, len(input))
+		for _, test := range input {
+			matched, regerr := regexp.MatchString(query.(string), test.Classifier["identifier"])
+			if matched && regerr == nil {
+				result = append(result, test)
+			}
+		}
+
+		return result
+	}
+
+	return input
+}
+
+func sendRequest(method string, path string, body interface{}, templateError string) ([]byte, error) {
+	var content []byte
+	if body != nil {
+		var err error
+		content, err = json.Marshal(body)
+		if err != nil {
+			return nil, fmt.Errorf(templateError, "unable to marshal the body in JSON.")
+		}
+	}
+
+	req, err := httputils.CreateRequest(method, path, content)
+	if err != nil {
+		return nil, fmt.Errorf(templateError, "unable to create request", err)
+	}
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf(templateError, "http request failure", err)
+	}
+
+	defer resp.Body.Close()
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf(templateError, "unable to read respond body", err)
+	}
+
+	if resp.StatusCode >= 300 {
```

Comment:
>300 [](start = 23, length = 3)

Sometimes is more clear with numbers, but in case you wanted something fancy, you can use [https://golang.org/pkg/net/http/#pkg-constants](https://golang.org/pkg/net/http/#pkg-constants)

Reply:
Yeah. I'm good. 300 is good...